bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/handler/LeapBridgeHandler.java

```diff
@@ -0,0 +1,802 @@
+                        logger.trace("Subject DN: {}", chain[cert].getSubjectDN());
+                        logger.trace("Issuer DN: {}", chain[cert].getIssuerDN());
+                        logger.trace("Serial number: {}", chain[cert].getSerialNumber());
+                    }
+                }
+            }
+
+            @Override
+            public X509Certificate @Nullable [] getAcceptedIssuers() {
+                return null;
+            }
+        } };
+    }
+
+    private synchronized void connect() {
+        deviceDataLoaded = false;
+        buttonDataLoaded = false;
+
+        try {
+            logger.debug("Opening SSL connection to {}:{}", config.ipAddress, config.port);
+            SSLSocket sslsocket = (SSLSocket) sslsocketfactory.createSocket(config.ipAddress, config.port);
+            sslsocket.startHandshake();
+            writer = new BufferedWriter(new OutputStreamWriter(sslsocket.getOutputStream()));
+            reader = new BufferedReader(new InputStreamReader(sslsocket.getInputStream()));
+            this.sslsocket = sslsocket;
+        } catch (UnknownHostException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Unknown host");
+            return;
+        } catch (IllegalArgumentException e) {
+            // port out of valid range
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Invalid port number");
+            return;
+        } catch (InterruptedIOException e) {
+            Thread.currentThread().interrupt();
+            logger.debug("Interrupted while establishing connection");
+            return;
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Error opening SSL connection. Check log.");
+            logger.info("Error opening SSL connection: {}", e.getMessage());
+            disconnect(false);
+            scheduleConnectRetry(reconnectInterval); // Possibly a temporary problem. Try again later.
+            return;
+        }
+
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, STATUS_INITIALIZING);
+
+        Thread readerThread = new Thread(this::readerThreadJob, "Lutron reader");
+        readerThread.setDaemon(true);
+        readerThread.start();
+        this.readerThread = readerThread;
+
+        Thread senderThread = new Thread(this::senderThreadJob, "Lutron sender");
+        senderThread.setDaemon(true);
+        senderThread.start();
+        this.senderThread = senderThread;
+
+        sendCommand(new LeapCommand(Request.getButtonGroups()));
+        queryDiscoveryData();
+        sendCommand(new LeapCommand(Request.subscribeOccupancyGroupStatus()));
+
+        logger.debug("Starting keepalive job with interval {}", heartbeatInterval);
+        keepAliveJob = scheduler.scheduleWithFixedDelay(this::sendKeepAlive, heartbeatInterval, heartbeatInterval,
+                TimeUnit.MINUTES);
```

Comment:
The spawning of asynchronous tasks inside of other asynchronous tasks can be a real source of problems if you are trying to dispose the handler while it is getting initialized. I suggest wrapping the Thread fields in a CompletableFuture so that you have a way to prevent starting them in the event that dispose it called before your asynchronous task initialized them.

```
CompletableFuture<Thread> readThreadFuture;

public void initialize() {
  readThreadFuture = new CompletableFuture<>();
  readThreadFuture.thenAccept(Thread::start);

  Thread readerThread = new Thread(this::readerThreadJob, "Lutron reader");
  readerThread.setDaemon(true);

  executor.submit(() -> {
    ...
    readThreadFuture.complete(readerThread);
  });
}

public void dispose(){
  //if this completes the future the asynchronous 'complete()' will be ignored
  if(!readThreadFuture.cancel(false)){
     //failure to cancel means that the future completed first, so lets interrupt the thread then   
     readThreadFuture.thenAccept(thread -> thread.interrupt());
  }
}
```

In addition could initialize your `keepAliveJob` inside of the `initialize()` method instead of doing it asynchronously? You could just make the keepAliveJob skip if the socket isn't ready yet.

Reply:
Sorry, I've been kind of tied up lately. I looked through that code again, though, and I don't really see any potential problems. Both connect() and disconnect() are synchronized. The dispose() routine calls cancel(true) on the Future running connect(), if it is still running, and then calls disconnect(). So disconnect() will wait for connect() to exit, and then clean up anything it did, including canceling connectRetryJob, keepAliveJob, and keepAliveReconnectJob, and interrupting senderThread and readerThread. It also closes the SSL socket, which would cause senderThread and readerThread to exit regardless. I should check to see if any of those should be declared as volatile to eliminate any potential inter-thread visibility problems, but other that than I don't really see any potential issues with dispose() whenever it's called. Very similar code is being used in this binding's ipbridge and in the alarm decoder binding's ipbridge.

Reply:
huh, I don't know why I didn't notice that connect and disconnect are synchronized. Yes that would prevent any problems so you can disregard my prior suggestion.